apollo-spaceport/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+    ))
+}
+
+impl ReportHeader {
+    fn try_new(graph: &str) -> Result<Self, ReporterError> {
+        let mut header = ReportHeader {
+            ..Default::default()
+        };
+
+        header.hostname = hostname()?;
+        header.agent_version = format!(
+            "{}-{}",
+            std::env!("CARGO_PKG_NAME"),
+            std::env!("CARGO_PKG_VERSION")
+        );
+        header.runtime_version = "rust".to_string();
+        header.uname = get_uname()?;
+        header.graph_ref = graph.to_string();
+        Ok(header)
+    }
+}
+
+/// The Reporter accepts requests from clients to transfer statistics
+/// and traces to the Apollo Ingress spaceport.
+#[derive(Debug)]
+pub struct Reporter {
+    client: ReporterClient<Channel>,
+}
+
+impl Reporter {
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new<T: AsRef<str>>(addr: T) -> Result<Self, ReporterError>
+    where
+        prost::bytes::Bytes: From<T>,
+    {
+        let ep = Endpoint::from_shared(addr)?;
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new_with_static(addr: &'static str) -> Result<Self, ReporterError> {
+        let ep = Endpoint::from_static(addr);
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Submit these stats onto the spaceport for eventual processing.
+    ///
+    /// The spaceport will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_stats(
+        &mut self,
+        graph: ReporterGraph,
+        key: String,
+        stats: ContextualizedStats,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        self.client
+            .add_stats(Request::new(ReporterStats {
+                graph: Some(graph),
+                key,
+                stats: Some(stats),
+            }))
+            .await
+    }
+
+    /// Submit this trace onto the spaceport for eventual processing.
+    ///
+    /// The spaceport will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_trace(
+        &mut self,
+        graph: ReporterGraph,
+        key: String,
+        trace: Trace,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        self.client
+            .add_trace(Request::new(ReporterTrace {
+                graph: Some(graph),
+                key,
+                trace: Some(trace),
+            }))
+            .await
+    }
+}
+
+/// The spaceport module contains the spaceport components
+pub mod spaceport {
```

Comment:
:yarn: Maybe move it to `spaceport.rs`?

Reply:
done: 6ff86de619bd6268ff41db6fce9da5f3d3dc22ea